corehq/util/couch_doc_processor.py

```diff
@@ -0,0 +1,323 @@
+        you may want to consider a different retry strategy.
+
+        :param doc: The doc dict to retry. It will be re-fetched from
+        the database before being yielded from the iteration.
+        :param max_retry: Maximum number of times a given document may
+        be retried.
+        :raises: `TooManyRetries` if this method has been called too
+        many times with a given document.
+        """
+        doc_id = doc["_id"]
+        retries = self.state["retry"].get(doc_id, 0) + 1
+        if retries > max_retry:
+            raise TooManyRetries(doc_id)
+        self.state["retry"][doc_id] = retries
+        self._save_state()
+
+    @property
+    def progress_info(self):
+        """Extra progress information
+
+        This property can be used to store and retrieve extra progress
+        information associated with the iteration. The information is
+        persisted with the iteration state in couch.
+        """
+        return self.state.get("progress_info")
+
+    @progress_info.setter
+    def progress_info(self, info):
+        self.state["progress_info"] = info
+        self._save_state()
+
+    def _save_state(self):
+        self.state["timestamp"] = datetime.utcnow().isoformat()
+        self.db.save_doc(self.state)
+
+    def discard_state(self):
+        try:
+            self.db.delete_doc(self.iteration_id)
+        except ResourceNotFound:
+            pass
+        self.__init__(
+            self.db,
+            self.original_doc_types,
+            self.iteration_key,
+            self.chunk_size,
+        )
+
+
+class ResumableDocsByTypeLogHandler(PaginateViewLogHandler):
+
+    def __init__(self, iterator):
+        self.iterator = iterator
+
+    def view_starting(self, db, view_name, kwargs, total_emitted):
+        offset = {k: v for k, v in kwargs.items() if k.startswith("startkey")}
+        self.iterator.state["offset"] = offset
+        self.iterator._save_state()
+
+
+class TooManyRetries(Exception):
+    pass
+
+
+DOCS_SKIPPED_WARNING = """
+        WARNING {} documents were not processed due to concurrent modification
+        during migration. Run the migration again until you do not see this
+        message.
+        """
+
+
+class BaseDocProcessor(six.with_metaclass(ABCMeta)):
+
+    def __init__(self, slug):
+        self.slug = slug
+
+    def __enter__(self):
+        pass
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        pass
+
+    @abstractmethod
+    def process_doc(self, doc, couchdb):
+        """Migrate a single document
+
+        :param doc: The document dict to be migrated.
+        :param couchdb: Couchdb database in which to save migrated doc.
+        :returns: True if doc was migrated else False. If this returns False
+        the document migration will be retried later.
+        """
+        raise NotImplementedError
+
+    def processing_complete(self, skipped):
+        pass
+
+    def filter(self, doc):
```

Comment:
suggestion: rename to `should_process`. It's not obvious from the name "filter" whether a return value of `True` means it should be omitted (to me this seems most logical from the name "filter") or processed.